src/lib.rs

```diff
@@ -175,6 +174,44 @@ impl From<TotpPublicKey> for PublicKey {
     }
 }
 
+impl TryFrom<RawPublicKey> for PublicKey {
+    type Error = serde::de::value::Error;
+
+    fn try_from(raw: RawPublicKey) -> Result<Self, Self::Error> {
+        match raw {
+            RawPublicKey {
+                kty: Some(Kty::Ec2),
+                alg: Some(Alg::Es256),
+                crv: Some(Crv::P256),
+                x: Some(x),
+                y: Some(y),
+            } => Ok(PublicKey::P256Key(P256PublicKey { x, y })),
+            RawPublicKey {
+                kty: Some(Kty::Ec2),
+                alg: Some(Alg::EcdhEsHkdf256),
+                crv: Some(Crv::P256),
+                x: Some(x),
+                y: Some(y),
+            } => Ok(PublicKey::EcdhEsHkdf256Key(EcdhEsHkdf256PublicKey { x, y })),
+            RawPublicKey {
+                kty: Some(Kty::Okp),
+                alg: Some(Alg::EdDsa),
+                crv: Some(Crv::Ed25519),
+                x: Some(x),
+                y: None,
+            } => Ok(PublicKey::Ed25519Key(Ed25519PublicKey { x })),
+            RawPublicKey {
+                kty: Some(Kty::Symmetric),
+                alg: Some(Alg::Totp),
+                crv: None,
+                x: None,
+                y: None,
+            } => Ok(PublicKey::TotpKey(TotpPublicKey {})),
+            _ => Err(serde::de::Error::custom("unknown key type")),
```

Comment:
This could also be a problem with `alg`/`crv`/`x`/`y` so I would prefer a less specific message, e. g.:
```suggestion
            _ => Err(serde::de::Error::custom("invalid key data")),
```